### PR TITLE
Fix listen dispose/unlisten memory leak

### PR DESCRIPTION
**Pull request by davidosomething**
Going to publish this into namespaced @sqs/alt and use that in site-server

- [x] Pending a SECSD ticket approval
    - https://jira.squarespace.net/servicedesk/customer/portal/38/SECSD-2287
- [ ] Add CODEOWNERS

### CODE REVIEW

**Comment by davidosomething** [comment]
@jpoo90 you already reviewed this against the goatslacker upstream
https://github.com/goatslacker/alt/pull/736

**Review by kauffecup — APPROVED** [review]

**Review by jacobangel — APPROVED** [review]

**Review by fringd — APPROVED** [review]

**Review by jpoo90 — APPROVED** [review]
@dotrakoun-sqsp do you foresee more alt changes ? I wonder if `alt` is something we want to keep improving.

**Comment by ghost** [comment]
> @dotrakoun-sqsp do you foresee more alt changes ? I wonder if `alt` is something we want to keep improving.

only if we find another bug

File: src/store/AltStore.js [hunk 1]
```diff
@@ -89,19 +89,26 @@ class AltStore {
...
     }
+    return () => this.unlisten(cb)
   }
 
   unlisten(cb) {
     this.lifecycle('unlisten')
-    this.subscriptions
-      .filter(subscription => subscription.cb === cb)
-      .forEach(subscription => subscription.dispose())
+    this.subscriptions = this.subscriptions.filter((s) => {
```
Comment (kelvin-d-zhu) [inline src/store/AltStore.js:105]: Should the object removed from `subscriptions` list too? filter does not remove it from the array.

Comment (kelvin-d-zhu) [reply src/store/AltStore.js:105]: nvm. i missed the `this.subscriptions = this.subscriptions.filte`  part

**Review by kelvin-d-zhu — APPROVED** [review]

File: src/store/AltStore.js [hunk 2]
```diff
@@ -89,19 +89,26 @@ class AltStore {
...
   listen(cb) {
     if (!fn.isFunction(cb)) throw new TypeError('listen expects a function')
-    const { dispose } = this.transmitter.subscribe(cb)
-    this.subscriptions.push({ cb, dispose })
-    return () => {
-      this.lifecycle('unlisten')
-      dispose()
+
+    // Already listening
+    const existing = this.subscriptions.find((s) => s.cb === cb)
```
Comment (istateside) [inline src/store/AltStore.js:94]: Isn't this a breaking change?

What do we expect to happen if we did something like this:
```
function listenHandler() { /* do something */ }

MyStore.listen(listenHandler);
MyStore.listen(listenHandler);
```

I would think we'd expect the callback to fire twice if we attached it manually twice.

What you've got does match the behavior of `addEventListener` though- adding the same callback to the same event multiple times doesn't seem to fire it more than once.

Comment (davidosomething) [reply src/store/AltStore.js:94]: Oh yea it is a different behavior from the original!


Based on the original unlisten code, it is also not intentional and I don't think we use it like this anywhere in our codebase.


Original listen: add A once, add A 2nd, and you get AA
Original unlisten: remove A once and you dispose() BOTH AA listeners -- the behavior is not symmetrical -- can't just dispose one or the other.


For our use case (and for sanity's sake) it is better to make this work like add/removeEventListener

**Comment by ghost** [comment]
@PyNate @swjclarke please take a look? looks like we need someone with write access to approve

**Comment by kelvin-d-zhu** [comment]
This is one of the causes of memory leak in v6 iframe. @PyNate @swjclarke, can you take a look?

**Review by swjclarke — APPROVED** [review]
I am approving as owner based on the other reviews. I do not have a strong understanding of the code here. 

Pleas increase the number of people who can in future provide owner reveiw

**Pull request merged**